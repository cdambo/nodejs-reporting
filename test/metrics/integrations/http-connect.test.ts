import { invoke, map, forEach, flatMap, constant } from "lodash";
import reporting, {
  requestTimeReporting,
  requestCountReporting,
  errorCountReporting,
  requestReporter,
  requestReporting,
  ReportingCreators,
  Handlers
} from "../../../src/metrics/integrations/http-connect";
import MetricsReporter from "../../../src/metrics/reporters/metrics-reporter";
import InMemoryReporter from "../../../src/metrics/reporters/in-memory-reporter";

const callsOfReporters = ({
  req = {},
  middlewareCreatorArgs = {},
  methods,
  handlerArgs = [{ thisIs: "a res" }, constant({})]
}: {
  req?: object;
  middlewareCreatorArgs?: object;
  methods: (keyof MetricsReporter)[];
  handlerArgs?: unknown[];
}): ReturnType<typeof jest.spyOn> => {
  const reporter = new InMemoryReporter();
  const mocks = map(
    methods,
    (method: keyof MetricsReporter): ReturnType<typeof jest.spyOn> =>
      jest.spyOn(reporter, method).mockImplementation()
  );
  const handlers = reporting({ reporter, ...middlewareCreatorArgs });
  forEach(
    handlers,
    (handler: (...args: unknown[]) => void): void =>
      handler(req, ...handlerArgs)
  );
  jest.runAllImmediates();
  return flatMap(
    mocks,
    (mock: ReturnType<typeof jest.spyOn>): ReturnType<typeof jest.spyOn> =>
      mock.mock.calls
  );
};

const callsOfReporter = ({
  creator,
  method,
  handlerArgs = [{ thisIs: "a req" }, { thisIs: "a res" }, constant({})]
}: {
  creator: (reporter: MetricsReporter) => (...args: unknown[]) => void;
  method: keyof MetricsReporter;
  handlerArgs?: unknown[];
}): ReturnType<typeof jest.spyOn> => {
  const reporter = new InMemoryReporter();
  const mockReporterMethod = jest.spyOn(reporter, method).mockImplementation();
  const handler = creator(reporter);
  handler(...handlerArgs);
  jest.runAllImmediates();
  return mockReporterMethod.mock.calls;
};

const reportsOf = ({
  middlewareCreator,
  middlewareCreatorArgs = {},
  method,
  handlerArgs = [{ thisIs: "a req" }, { thisIs: "a res" }, constant({})]
}: {
  middlewareCreator: (...args: unknown[]) => (...args: unknown[]) => void;
  middlewareCreatorArgs?: object;
  method: keyof MetricsReporter;
  handlerArgs?: unknown[];
}): ReturnType<typeof jest.spyOn> => {
  const creator = (reporter: MetricsReporter): ((...args: unknown[]) => void) =>
    middlewareCreator({ reporter, ...middlewareCreatorArgs });

  return callsOfReporter({ creator, method, handlerArgs });
};

describe("Http Connect", (): void => {
  it("Creates a list of reporting handlers", (): void => {
    const req = {};
    expect(
      callsOfReporters({ req, methods: ["increment", "timing"] })
    ).toMatchSnapshot();
    expect(req).toMatchSnapshot();
  });

  it("Overrides specific handler configuration", (): void => {
    expect(
      callsOfReporters({
        methods: ["increment", "timing"],
        middlewareCreatorArgs: {
          handlers: [Handlers.RequestCount, Handlers.RequestTime],
          getTags: constant({ thisIs: "a_tag" }),
          sampleRate: 0.6,
          timeReporting: {
            getTags: constant({ thisIsNot: "the_same_tag" }),
            sampleRate: 0.8,
            stat: "what_a_stat"
          },
          countReporting: {
            stat: "what_another_stat"
          }
        }
      })
    ).toMatchSnapshot();
  });

  describe("requestTimeReporting", (): void => {
    it("Reports a time metric", (): void => {
      expect(
        reportsOf({
          middlewareCreator: requestTimeReporting,
          method: "timing",
          handlerArgs: [{ thisIs: "a req" }, { thisIs: "a res" }, 18]
        })
      ).toMatchSnapshot();
    });

    it("Reports a time metric with a stat name", (): void => {
      expect(
        reportsOf({
          middlewareCreator: requestTimeReporting,
          method: "timing",
          middlewareCreatorArgs: { stat: "time_stat" },
          handlerArgs: [{ thisIs: "a req" }, { thisIs: "a res" }, 18]
        })
      ).toMatchSnapshot();
    });

    it("Reports a time metric with a sample rate", (): void => {
      expect(
        reportsOf({
          middlewareCreator: requestTimeReporting,
          method: "timing",
          middlewareCreatorArgs: { sampleRate: 0.5 },
          handlerArgs: [{ thisIs: "a req" }, { thisIs: "a res" }, 18]
        })
      ).toMatchSnapshot();
    });

    it("Reports a time metric with a getTags function", (): void => {
      expect(
        reportsOf({
          middlewareCreator: requestTimeReporting,
          method: "timing",
          middlewareCreatorArgs: { getTags: constant({ tag: "one" }) },
          handlerArgs: [{ thisIs: "a req" }, { thisIs: "a res" }, 18]
        })
      ).toMatchSnapshot();
    });
  });

  describe("requestCountReporting", (): void => {
    it("Reports a count metric", (): void => {
      expect(
        reportsOf({
          middlewareCreator: requestCountReporting,
          method: "increment"
        })
      ).toMatchSnapshot();
    });

    it("Reports a count metric with a stat name", (): void => {
      expect(
        reportsOf({
          middlewareCreator: requestCountReporting,
          method: "increment",
          middlewareCreatorArgs: { stat: "count_stat" }
        })
      ).toMatchSnapshot();
    });

    it("Reports a count metric with a sample rate", (): void => {
      expect(
        reportsOf({
          middlewareCreator: requestCountReporting,
          method: "increment",
          middlewareCreatorArgs: { sampleRate: 0.5 }
        })
      ).toMatchSnapshot();
    });

    it("Reports a count metric with a getTags function", (): void => {
      expect(
        reportsOf({
          middlewareCreator: requestCountReporting,
          method: "increment",
          middlewareCreatorArgs: { getTags: constant({ tag: "one" }) }
        })
      ).toMatchSnapshot();
    });
  });

  describe("errorCountReporting", (): void => {
    const handlerArgs = [
      new Error(),
      { thisIs: "a req" },
      { thisIs: "a res" },
      (err: Error): Error => err
    ];

    it("Reports a count metric", (): void => {
      expect(
        reportsOf({
          middlewareCreator: errorCountReporting,
          method: "increment",
          handlerArgs
        })
      ).toMatchSnapshot();
    });

    it("Reports a count metric with a stat name", (): void => {
      expect(
        reportsOf({
          middlewareCreator: errorCountReporting,
          method: "increment",
          middlewareCreatorArgs: { stat: "error_count_stat" },
          handlerArgs
        })
      ).toMatchSnapshot();
    });

    it("Reports a count metric with a sample rate", (): void => {
      expect(
        reportsOf({
          middlewareCreator: errorCountReporting,
          method: "increment",
          middlewareCreatorArgs: { sampleRate: 0.5 },
          handlerArgs
        })
      ).toMatchSnapshot();
    });

    it("Reports a count metric with a getTags function", (): void => {
      expect(
        reportsOf({
          middlewareCreator: errorCountReporting,
          method: "increment",
          middlewareCreatorArgs: { getTags: constant({ tag: "one" }) },
          handlerArgs
        })
      ).toMatchSnapshot();
    });
  });

  describe("requestReporter", (): void => {
    it("Adds the reporter to the req object", (): void => {
      const reporter = new InMemoryReporter();
      const req = {};
      requestReporter({ reporter })(req, { thisIs: "a res" }, constant({}));
      expect(req).toMatchSnapshot();
    });
  });

  describe("requestReporting", (): void => {
    const reportsOfRequestReporting = ({
      creatorMethod,
      creatorArgs = {},
      method,
      handlerArgs = [{ thisIs: "a req" }, { thisIs: "a res" }, constant({})]
    }: {
      creatorMethod: keyof ReportingCreators;
      creatorArgs?: object;
      method: keyof MetricsReporter;
      handlerArgs?: unknown[];
    }): ReturnType<typeof jest.spyOn> => {
      const creator = (
        reporter: MetricsReporter
      ): ((...args: unknown[]) => void) =>
        invoke(requestReporting({ reporter }), creatorMethod, creatorArgs);

      return callsOfReporter({ creator, method, handlerArgs });
    };

    describe("timing", (): void => {
      it("Returns a ResponseTimeHandler", (): void => {
        expect(
          reportsOfRequestReporting({
            creatorMethod: "timing",
            creatorArgs: {
              stat: "request-time-reporting",
              sampleRate: 0.5,
              getTags: constant({ one: "tag" })
            },
            method: "timing",
            handlerArgs: [{ thisIs: "a req" }, { thisIs: "a res" }, 18]
          })
        ).toMatchSnapshot();
      });
    });

    describe("count", (): void => {
      it("Returns a RequestHandler", (): void => {
        expect(
          reportsOfRequestReporting({
            creatorMethod: "count",
            creatorArgs: {
              stat: "request-count-reporting",
              sampleRate: 0.5,
              getTags: constant({ one: "tag" })
            },
            method: "increment"
          })
        ).toMatchSnapshot();
      });
    });

    describe("errors", (): void => {
      it("Returns an ErrorHandler", (): void => {
        expect(
          reportsOfRequestReporting({
            creatorMethod: "errors",
            creatorArgs: {
              stat: "request-time-reporting",
              sampleRate: 0.5,
              getTags: constant({ one: "tag" })
            },
            method: "increment",
            handlerArgs: [
              new Error(),
              { thisIs: "a req" },
              { thisIs: "a res" },
              (err: Error): Error => err
            ]
          })
        ).toMatchSnapshot();
      });
    });

    describe("reqReporter", (): void => {
      it("Adds the reporter to the req object", (): void => {
        const reporter = new InMemoryReporter();
        const req = {};
        requestReporting({ reporter }).reqReporter()(
          req,
          { thisIs: "a res" },
          constant({})
        );
        expect(req).toMatchSnapshot();
      });
    });
  });
});
