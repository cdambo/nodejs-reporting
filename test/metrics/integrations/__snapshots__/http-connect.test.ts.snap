// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Http Connect Creates a list of reporting handlers 1`] = `
Array [
  Array [
    "http_requests_total",
    1,
    Object {},
  ],
  Array [
    "http_request_duration_seconds",
    18,
    1,
    Object {},
  ],
]
`;

exports[`Http Connect Creates a list of reporting handlers 2`] = `
Object {
  "reporter": InMemoryReporter {
    "context": Context {
      "globalTags": Object {},
    },
    "format": Object {
      "format": [Function],
    },
    "increment": [MockFunction] {
      "calls": Array [
        Array [
          "http_requests_total",
          1,
          Object {},
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    },
    "reports": Array [],
    "timing": [MockFunction] {
      "calls": Array [
        Array [
          "http_request_duration_seconds",
          18,
          1,
          Object {},
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    },
  },
  "thisIs": "a request",
}
`;

exports[`Http Connect Overrides specific handler configuration 1`] = `
Array [
  Array [
    "what_another_stat",
    0.6,
    Object {
      "thisIs": "a_tag",
    },
  ],
  Array [
    "what_a_stat",
    20,
    0.8,
    Object {
      "thisIsNot": "the_same_tag",
    },
  ],
]
`;

exports[`Http Connect errorCountReporting Reports a count metric 1`] = `
Array [
  Array [
    "http_requests_errors_total",
    1,
    Object {},
  ],
]
`;

exports[`Http Connect errorCountReporting Reports a count metric with a getTags function 1`] = `
Array [
  Array [
    "http_requests_errors_total",
    1,
    Object {
      "tag": "one",
    },
  ],
]
`;

exports[`Http Connect errorCountReporting Reports a count metric with a sample rate 1`] = `
Array [
  Array [
    "http_requests_errors_total",
    0.5,
    Object {},
  ],
]
`;

exports[`Http Connect errorCountReporting Reports a count metric with a stat name 1`] = `
Array [
  Array [
    "error_count_stat",
    1,
    Object {},
  ],
]
`;

exports[`Http Connect requestCountReporting Reports a count metric 1`] = `
Array [
  Array [
    "http_requests_total",
    1,
    Object {},
  ],
]
`;

exports[`Http Connect requestCountReporting Reports a count metric with a getTags function 1`] = `
Array [
  Array [
    "http_requests_total",
    1,
    Object {
      "tag": "one",
    },
  ],
]
`;

exports[`Http Connect requestCountReporting Reports a count metric with a sample rate 1`] = `
Array [
  Array [
    "http_requests_total",
    0.5,
    Object {},
  ],
]
`;

exports[`Http Connect requestCountReporting Reports a count metric with a stat name 1`] = `
Array [
  Array [
    "count_stat",
    1,
    Object {},
  ],
]
`;

exports[`Http Connect requestReporter Adds the reporter to the req object 1`] = `
Object {
  "reporter": InMemoryReporter {
    "context": Context {
      "globalTags": Object {},
    },
    "format": Object {
      "format": [Function],
    },
    "reports": Array [],
  },
  "thisIs": "a request",
}
`;

exports[`Http Connect requestReporters count Returns a RequestHandler 1`] = `
Array [
  Array [
    "request-count-reporting",
    0.5,
    Object {
      "one": "tag",
    },
  ],
]
`;

exports[`Http Connect requestReporters errors Returns an ErrorHandler 1`] = `
Array [
  Array [
    "request-time-reporting",
    0.5,
    Object {
      "one": "tag",
    },
  ],
]
`;

exports[`Http Connect requestReporters reqReporter Adds the reporter to the req object 1`] = `
Object {
  "reporter": InMemoryReporter {
    "context": Context {
      "globalTags": Object {},
    },
    "format": Object {
      "format": [Function],
    },
    "reports": Array [],
  },
  "thisIs": "a request",
}
`;

exports[`Http Connect requestReporters timing Returns a ResponseTimeHandler 1`] = `
Array [
  Array [
    "request-time-reporting",
    18,
    0.5,
    Object {
      "one": "tag",
    },
  ],
]
`;

exports[`Http Connect requestTimeReporting Reports a time metric 1`] = `
Array [
  Array [
    "http_request_duration_seconds",
    18,
    1,
    Object {},
  ],
]
`;

exports[`Http Connect requestTimeReporting Reports a time metric with a getTags function 1`] = `
Array [
  Array [
    "http_request_duration_seconds",
    18,
    1,
    Object {
      "tag": "one",
    },
  ],
]
`;

exports[`Http Connect requestTimeReporting Reports a time metric with a sample rate 1`] = `
Array [
  Array [
    "http_request_duration_seconds",
    18,
    0.5,
    Object {},
  ],
]
`;

exports[`Http Connect requestTimeReporting Reports a time metric with a stat name 1`] = `
Array [
  Array [
    "time_stat",
    18,
    1,
    Object {},
  ],
]
`;
